### ISSUE TITLE
plot_source messes up calc_energy_flux/plot_fit/...probably others... for PHA data sets

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#673]
In the following, the original `plot_fit` call is "correct", but then a call to `plot_source` makes subsequent calls to `plot_fit` to artificially increase the model values (by a large amount). I haven't included the output images, but you can see the y range increases dramatically:

|           | minimum | maximum |
| --- | --- | --- |
| first | -1.4e-4 | 2.7e-3 |
| second | -2.6e-4 | 5.4e-2 |

```
sherpa In [1]: %matplotlib
Using matplotlib backend: TkAgg

sherpa In [2]: load_pha('sherpa-test-data/sherpatest/obs1.pi')
read ARF file sherpa-test-data/sherpatest/obs1.arf
read RMF file sherpa-test-data/sherpatest/obs1.rmf
read background file sherpa-test-data/sherpatest/obs1_bg.pi

sherpa In [3]: group_counts(5)

sherpa In [4]: set_source(powlaw1d.pl)

sherpa In [5]: pl.gamma = 1.5

sherpa In [6]: pl.ampl = 9.5e-7

sherpa In [7]: plot_fit()

sherpa In [8]: from matplotlib import pyplot as plt

sherpa In [9]: plt.ylim()
Out[9]: (-0.00013627224824930666, 0.0027496860510178647)

sherpa In [10]: plt_source()
NameError: name 'plt_source' is not defined

sherpa In [11]: plot_source()

sherpa In [12]: plot_fit()

sherpa In [13]: plt.ylim()
Out[13]: (-0.0025796359269162563, 0.054060323303023801)

```

**Comment by DougBurke** [comment]
It also happens with chips.

**Comment by DougBurke** [comment]
Repeated calls to

    plot_source()
    plot_fit()

increase the "amplitude" of both the source and fit plots. Is this something to do with caching (e.g. the amplitude is being applied to the cached result and then stored again): **complete** guess.

**Comment by DougBurke** [comment]
See how the output of `calc_energy_flux` increases each time `plot_source` is called. It is by a factor of 100, which is 1 / bin-width-in-kev for this dataset (may be a coincidence).

```
sherpa In [1]: load_pha('/lagado.real/sherpa/sherpa-master/sherpa-test-data/sher
          ...: patest/obs1.pi')
read ARF file /lagado.real/sherpa/sherpa-master/sherpa-test-data/sherpatest/obs1.arf
read RMF file /lagado.real/sherpa/sherpa-master/sherpa-test-data/sherpatest/obs1.rmf
read background file /lagado.real/sherpa/sherpa-master/sherpa-test-data/sherpatest/obs1_bg.pi

sherpa In [2]: set_source(powlaw1d.pl)

sherpa In [3]: pl.gamma = 1.5

sherpa In [4]: pl.ampl = 9.5e-7

sherpa In [5]: calc_energy_flux(lo=0.5, hi=7)
Out[5]: 5.9072825997916164e-15

sherpa In [6]: calc_energy_flux(lo=0.5, hi=7)
Out[6]: 5.9072825997916164e-15

sherpa In [7]: plot_source()

sherpa In [8]: calc_energy_flux(lo=0.5, hi=7)
Out[8]: 5.9072824691596447e-13

sherpa In [9]: calc_energy_flux(lo=0.5, hi=7)
Out[9]: 5.9072824691596447e-13

sherpa In [10]: plot_source()

sherpa In [11]: calc_energy_flux(lo=0.5, hi=7)
Out[11]: 5.9072823397026535e-11

```

**Comment by DougBurke** [comment]
For the run above, we can see that the model amplitude isn't being changed, so I think it must be something internal to Sherpa (which is why I think it might be some of the cache-related code):

```
sherpa In [13]: print(pl)
powlaw1d.pl
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   pl.gamma     thawed          1.5          -10           10           
   pl.ref       frozen            1 -3.40282e+38  3.40282e+38           
   pl.ampl      thawed      9.5e-07            0  3.40282e+38           

sherpa In [14]: calc_energy_flux(lo=0.5, hi=7)
Out[14]: 5.9072823397026535e-11

sherpa In [15]: print(pl)
powlaw1d.pl
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   pl.gamma     thawed          1.5          -10           10           
   pl.ref       frozen            1 -3.40282e+38  3.40282e+38           
   pl.ampl      thawed      9.5e-07            0  3.40282e+38           

sherpa In [16]: plot_source()

sherpa In [17]: calc_energy_flux(lo=0.5, hi=7)
Out[17]: 5.9072822114206463e-09

sherpa In [18]: print(pl)
powlaw1d.pl
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   pl.gamma     thawed          1.5          -10           10           
   pl.ref       frozen            1 -3.40282e+38  3.40282e+38           
   pl.ampl      thawed      9.5e-07            0  3.40282e+38           

```

The `calc_energy_flux` output increases with the call to `plot_source` but the model itself hasn't changed (and neither does the output of `get_source` or `get_model`):

```
sherpa In [19]: get_source()
Out[19]: <PowLaw1D model instance 'powlaw1d.pl'>

sherpa In [20]: get_model()
Out[20]: <RSPModelPHA model instance 'apply_rmf(apply_arf((43933.416425609 * powlaw1d.pl)))'>

sherpa In [21]: get_exptime()
NameError: name 'get_exptime' is not defined

sherpa In [22]: get_exposure()
Out[22]: 43933.416425609001

```

**Comment by DougBurke** [comment]
I created a little script which doesn't rely on any data file (i.e it fakes the pha, arf, and rmf) to show the problem, and that both `plot_source` and `plot_model` are affected. The differences show that the values are being increased by 1 / bin width (since the example uses a constant bin width).

https://gist.github.com/DougBurke/3cd22224e633c46c508fb0c5c2cbcf26

**Comment by DougBurke** [comment]
In this script, although both model and source report differences, it is because of the first `plot_source` call - if you comment out the `s1y` and `s2y` lines then the `m1y` and `m2y` values are the same.

I've tried to track down the logic of what is going on, but can not replicate it - e.g. with

```
mdl = ui._session._sources[1]
dat = ui.get_data(1)
vals1 = dat.to_plot(yfunc=mdl)
vals2 = dat.to_plot(yfunc=mdl)
```

which is basically what sherpa.plot.ModelPlot.prepare` is doing, but do not see anything odd. So I'm obviously missing something.

**Comment by wmclaugh** [comment]
Note that this appears to be a regression which was introduced in CIAO 4.11.

Using sherpa from the Linux version of CIAO 4.8, 4.9, and 4.10 all yield the same result (no shift in the y axis range on repeated calls to get_plot_source):

    -----------------------------------------------------
    Welcome to Sherpa: CXC's Modeling and Fitting Package
    -----------------------------------------------------
    CIAO 4.8 Sherpa version 1 Thursday, December 3, 2015

    sherpa-1> load_pha('/export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.pi')
    set_source(powlaw1d.pl)
    pl.gamma = 1.5
    pl.ampl = 9.5e-7
    read ARF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.arf
    read RMF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.rmf
    read background file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1_bg.pi
    sherpa-2> get_source_plot().y[0:5]
    array([  8.90397428e-06,   8.34151080e-06,   7.83587607e-06,
             7.37934344e-06,   6.96548626e-06])
    sherpa-3> get_source_plot().y[0:5]
    array([  8.90397428e-06,   8.34151080e-06,   7.83587607e-06,
             7.37934344e-06,   6.96548626e-06])
    sherpa-4> get_source_plot().y[0:5]
    array([  8.90397428e-06,   8.34151080e-06,   7.83587607e-06,
             7.37934344e-06,   6.96548626e-06])
    sherpa-5> get_source_plot().y[0:5]
    array([  8.90397428e-06,   8.34151080e-06,   7.83587607e-06,
             7.37934344e-06,   6.96548626e-06])

With the CIAO 4.11 version of sherpa, we see the shift:
    -----------------------------------------------------
    Welcome to Sherpa: CXC's Modeling and Fitting Package
    -----------------------------------------------------
    CIAO 4.11 Sherpa version 1 Wednesday, December 5, 2018

    Python 3.5.4 (default, Oct 15 2018, 13:47:46)
    Type 'copyright', 'credits' or 'license' for more information
    IPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.
    IPython profile: sherpa
    sherpa In [1]: load_pha('/export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1
              ...: .pi')
              ...: set_source(powlaw1d.pl)
              ...: pl.gamma = 1.5
              ...: pl.ampl = 9.5e-7
    read ARF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.arf
    read RMF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.rmf
    read background file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1_bg.pi
    sherpa In [2]: get_source_plot().y[0:5]
    Out[2]:
    array([  8.90397428e-06,   8.34151080e-06,   7.83587607e-06,
             7.37934344e-06,   6.96548626e-06])
    sherpa In [3]: get_source_plot().y[0:5]
    Out[3]: array([ 0.0008904 ,  0.00083415,  0.00078359,  0.00073794,  0.00069655])
    sherpa In [4]: get_source_plot().y[0:5]
    Out[4]: array([ 0.08903965,  0.08341527,  0.07835868,  0.07379358,  0.06965458])

**Comment by wmclaugh** [comment]
Using the git bisect command and the following test case:

    from sherpa.astro.ui import *
    import numpy as np
    from sherpa import *
    from sherpa.ui import *
    from sherpa.astro import *
    from sherpa.astro.ui import *

    load_pha('/export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.pi')
    set_source(powlaw1d.pl)
    pl.gamma = 1.5
    pl.ampl = 9.5e-7
    print(get_source_plot().y[0:5])
    print(get_source_plot().y[0:5])
    print(get_source_plot().y[0:5])

I've gone through the commits since CIAO 4.10 and it appears that the regression was introduced in:

commit a0999bb5af320266231528b7756f055806b56e02
Date:   Wed Apr 18 13:07:26 2018 -0400

    Use cached models when possible

:040000 040000 4cc05fff2bc5fa0be546c4b9387a0add505fbfba 2544fc4863781122c743d95dc803789aea6d9b3b M	sherpa

Commits prior to that commit return:

    (base) devel14-121: python /pool7/wlm/test_source2.py
    WARNING: imaging routines will not be available, 
    failed to import sherpa.image.ds9_backend due to 
    'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
    WARNING: failed to import sherpa.plot.pylab_backend; plotting routines will not be available
    WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
    read ARF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.arf
    read RMF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.rmf
    read background file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1_bg.pi
    [8.90397428e-06 8.34151080e-06 7.83587607e-06 7.37934344e-06
     6.96548626e-06]
    [8.90397428e-06 8.34151080e-06 7.83587607e-06 7.37934344e-06
     6.96548626e-06]
    [8.90397428e-06 8.34151080e-06 7.83587607e-06 7.37934344e-06
     6.96548626e-06]

while subsequent commits exhibit the change in Y range:

    (base) devel14-118: python /pool7/wlm/test_source2.py
    WARNING: imaging routines will not be available, 
    failed to import sherpa.image.ds9_backend due to 
    'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
    WARNING: failed to import sherpa.plot.pylab_backend; plotting routines will not be available
    WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
    read ARF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.arf
    read RMF file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1.rmf
    read background file /export/testdata_c45/testdata/ciao4.3/setfullmodel/obs1_bg.pi
    [8.90397428e-06 8.34151080e-06 7.83587607e-06 7.37934344e-06
     6.96548626e-06]
    [0.0008904  0.00083415 0.00078359 0.00073794 0.00069655]
    [0.08903965 0.08341527 0.07835868 0.07379358 0.06965458]

**Comment by dtnguyen2** [comment]
After initially denying that my PR of the cache code is the source of the problem reported here, I was _asked_ to look into the bug report anyway.  It turns out, at least I think, the source of the problem is in the ```modelCacher1d(func)``` code which I did not touch in my PR.  One of the changes in my PR code is the following modification in the models/model.py file, it was changed from:
```
        self._use_caching = False  # FIXME: reduce number of variables?
```
to
```
        self._use_caching = True  # FIXME: reduce number of variables?
```

That is my PR turned the member ```_use_caching``` from ```False``` to ```True``` then caching will be used in the ```modelCacher1d(func)``` code, see:

```
def modelCacher1d(func):

    def cache_model(cls, pars, xlo, *args, **kwargs):
        use_caching = cls._use_caching
```

The caching code relies on several keys: 1) the fitted pars, 2) whether data is to be integrated or not, 3) xlo, 4) args.  I modified the ```modelCacher1d(func)``` to print out the keys and the 'y' value when they are calculated or using the cached values as follows (ie only print the first and last element of the array):

```
(sherpa) [dtn@devel12 sherpa]$ git diff models/model.py
diff --git a/sherpa/models/model.py b/sherpa/models/model.py
index bff47548..c6917bb0 100644
--- a/sherpa/models/model.py
+++ b/sherpa/models/model.py
@@ -54,7 +54,12 @@ def modelCacher1d(func):
 
         digest = ''
         if use_caching:
-
+            print('pars =', numpy.array(pars))
+            print('kwargs.get(integrate) =', kwargs.get('integrate'))
+            nxlo = len(xlo)
+            print('xlo =', nxlo, xlo[0], xlo[nxlo-1])
+            nargs0 = len(args[0])
+            print('args =', nargs0, args[0][0], args[0][nargs0-1])
             data = [numpy.array(pars).tostring(), boolean_to_byte(kwargs.get('integrate', False)),
                     numpy.asarray(xlo).tostring()]
             if args:
@@ -63,9 +68,14 @@ def modelCacher1d(func):
             token = b''.join(data)
             digest = hashlib.sha256(token).digest()
             if digest in cache:
+                tmp = cache[digest]
+                print('using cache =', tmp[0], tmp[len(tmp)-1])
+                print()
                 return cache[digest]
 
         vals = func(cls, pars, xlo, *args, **kwargs)
+        print('will cache =', vals[0], vals[len(vals)-1])
+        print()
 
         if use_caching:
             # remove first item in queue and remove from cache
```

So by running Doug's script, I get the following output:

```
(sherpa) [dtn@devel12 sherpa]$ python tmp.py 
pars = [1.5e+00 1.0e+00 9.5e-07]
kwargs.get(integrate) = None
xlo = 1078 0.2199999988079071 10.989999771118164
args = 1078 0.23000000417232513 11.0
will cache = 8.903979058142175e-08 2.605798068044618e-10

pars = [1.5e+00 1.0e+00 9.5e-07]
kwargs.get(integrate) = None
xlo = 1078 0.2199999988079071 10.989999771118164
args = 1078 0.23000000417232513 11.0
using cache = 8.903979058142175e-08 2.605798068044618e-10

pars = [1.5e+00 1.0e+00 9.5e-07]
kwargs.get(integrate) = None
xlo = 1078 0.2199999988079071 10.989999771118164
args = 1078 0.23000000417232513 11.0
using cache = 8.903974281678157e-06 2.6057384274250938e-08
```

As one can see, the keys (pars, integrate, xlo and args) do not change in the three calls within ```modelCacher1d(func)```. The first entry into the function it will cache the following values
(printing only the first and last values of the y-array):

```
will cache = 8.903979058142175e-08 2.605798068044618e-10
```
The 2nd entry into the ```modelCacher1d(func)``` it uses the following cached values:
```
using cache = 8.903979058142175e-08 2.605798068044618e-10
```
which was calculated initially. However, in the 3rd entry into the ```modelCacher1d(func)```, the returned cached values are off by a factor of 100:
```
using cache = 8.903974281678157e-06 2.6057384274250938e-08
```

I guess, whoever coded the ```modelCacher1d(func)``` must have realized there is an issue and turn off the caching. I think I have an **ugly** solution to the mess but I am still investigating a better one.

**Comment by DougBurke** [comment]
So, my assumption is that the model evaluation is being cached, and then something somewhere takes the array from the cache and modifies it in place, hence changing the cached value. To test this assumption out I made the cached value immutable - by saying in `sherpa/models/model.py::modelCache1d::cache_model`

```
% git diff sherpa/models/model.py 
diff --git a/sherpa/models/model.py b/sherpa/models/model.py
index a0c3129..46d703e 100644
--- a/sherpa/models/model.py
+++ b/sherpa/models/model.py
@@ -76,6 +76,14 @@ def modelCacher1d(func):
             queue.append(digest)
             cache[digest] = vals
 
+            # make sure the cached value is immutable; perhaps should
+            # then return a copy of the array that is not flagged
+            # this way (to avoid bizarre behavior hidden at a distance
+            # or behind a cache flag)
+            #
+            cache[digest].setflags(write=False)
+            # note: this changes vals too
+
         return vals
 
     cache_model.__name__ = func.__name__
```

and then my script in https://gist.github.com/DougBurke/3cd22224e633c46c508fb0c5c2cbcf26 fails with

```
python /pool14/dburke/sherpa-bug.py 
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
Traceback (most recent call last):
  File "/pool14/dburke/sherpa-bug.py", line 36, in <module>
    s1y = ui.get_source_plot().y.copy()
  File "<string>", line 1, in get_source_plot
  File "/lagado.real/sherpa/sherpa-plotting/sherpa/astro/ui/utils.py", line 10139, in get_source_plot
    return self._prepare_plotobj(id, self._astrosourceplot, lo=lo, hi=hi)
  File "/lagado.real/sherpa/sherpa-plotting/sherpa/astro/ui/utils.py", line 10939, in _prepare_plotobj
    plotobj.prepare(data, src, lo, hi)
  File "/lagado.real/sherpa/sherpa-plotting/sherpa/astro/plot.py", line 167, in prepare
    self.y /= xmid
ValueError: output array is read-only
```

This lead me to the following patch, which fixes this particular problem, but doesn't solve the overall issue (of accidentally corrupting the cache).

```
git diff sherpa/astro/plot.py 
diff --git a/sherpa/astro/plot.py b/sherpa/astro/plot.py
index 693c3f0..5ea9d7a 100644
--- a/sherpa/astro/plot.py
+++ b/sherpa/astro/plot.py
@@ -143,7 +143,8 @@ class SourcePlot(HistogramPlot):
         self.title  = 'Source Model of %s' % data.name
         self.xlo, self.xhi = data._get_indep(filter=False)
         self.mask = filter_bins((lo,), (hi,), (self.xlo,))
-        self.y = src(self.xlo, self.xhi)
+        self.y = src(self.xlo, self.xhi).copy()
+
         prefix_quant = 'E'
         quant = 'keV'
 
```

An alternative would be for the model cache routine to return a copy of the cached data rather than the data itself - i.e.

```
% git diff sherpa/models/model.py 
diff --git a/sherpa/models/model.py b/sherpa/models/model.py
index a0c3129..3f599a1 100644
--- a/sherpa/models/model.py
+++ b/sherpa/models/model.py
@@ -52,6 +52,12 @@ def modelCacher1d(func):
         cache = cls._cache
         queue = cls._queue
 
+        # Note: we want to make sure the user doesn't accidentally
+        # change the cached values, so we make sure the cached copy
+        # is not the same as returned to the user. See
+        # https://github.com/sherpa/sherpa/issues/673
+        #
+
         digest = ''
         if use_caching:
 
@@ -63,7 +69,7 @@ def modelCacher1d(func):
             token = b''.join(data)
             digest = hashlib.sha256(token).digest()
             if digest in cache:
-                return cache[digest]
+                return cache[digest].copy()
 
         vals = func(cls, pars, xlo, *args, **kwargs)
 
@@ -74,7 +80,7 @@ def modelCacher1d(func):
 
             # append newest model values to queue
             queue.append(digest)
-            cache[digest] = vals
+            cache[digest] = vals.copy()
 
         return vals
 
```

which I think is safer.

**Comment by DougBurke** [comment]
Interestingly enough (at least to me) - if I add the `.setFlags(write=False)` to the cache and then run the tests (without XSPEC available) then only one test fails.

**Comment by DougBurke** [comment]
This has been fixed by #705
